### PR TITLE
custom comp: do not complete flags after args when interspersed is false

### LINKED ISSUE #1307: custom completion should respect the flagset interspersed option
The flagset interspersed option indicates that flags which follow after the fist argument are also considered arguments and are not parsed as flag.

Consider this simple example:
```
package main

import (
	"fmt"

	"github.com/spf13/cobra"
)

var rootCmd = &cobra.Command{
	Use: "testprog",
	Run: func(cmd *cobra.Command, args []string) {
		fmt.Println("rootCmd called")
	},
}

var subCmd = &cobra.Command{
	Use: "sub",
	Run: func(cmd *cobra.Command, args []string) {
		fmt.Println(args)
	},
}

func main() {
	rootCmd.AddCommand(subCmd)
	flags := subCmd.Flags()
	flags.SetInterspersed(false)
	flags.Bool("test", false, "test flag")

	rootCmd.Execute()
}
```

The --test flag is added to args if an argument exists before:
```
$ ./testprog sub --test a
[a]
$ ./testprog sub a --test 
[a --test]
```

This is the expected behaviour however the custom completion logic does not respects this. It still provides flag completion even after the first arg is set.

```
$ ./testprog __complete sub a --t
--test  test flag
:4
Completion ended with directive: ShellCompDirectiveNoFileComp
```
This should not happen. Instead it should provide argument completion (e.g. ValidArgsFunction).

I looked at the code but I'm not sure how this could be implemented. The problem is that the interspersed option is not exported from the FlagSet so we cannot retrieve it.

@marckhouzam Any Ideas?

### PR DESCRIPTION
**Pull request by Luap99**
If the interspersed option is set false and one arg is already set all
following arguments are counted as arg and not parsed as flags. Because
of that we should not offer flag completion. The same applies to arguments
followed after `--`.

Fixes #1307

### CODE REVIEW

File: custom_completions_test.go [hunk 1]
```diff
@@ -1697,6 +1697,107 @@ func TestValidArgsFuncChildCmdsWithDesc(t *testing.T) {
...
+
+	if output != expected {
+		t.Errorf("expected: %q, got: %q", expected, output)
+	}
+
+	// Test that no flags are completed after the -- arg with a flag set
+	output, err = executeCommand(rootCmd, ShellCompRequestCmd, "--bool", "--", "-")
+	if err != nil {
+		t.Errorf("Unexpected error: %v", err)
+	}
```
Comment (marckhouzam) [inline custom_completions_test.go:1743]: I think we need one more test:
```
output, err = executeCommand(rootCmd, ShellCompRequestCmd, "--", "--string", "")
```
to make sure that we don't get flag value completion but get `ValidArgsFunction` instead.

File: custom_completions.go [hunk 2]
```diff
@@ -284,7 +298,7 @@ func (c *Command) getCompletions(args []string) (*Command, []string, ShellCompDi
 
 	var completions []string
 	directive := ShellCompDirectiveDefault
-	if flag == nil {
+	if flag == nil && flagCompletion {
```
Comment (marckhouzam) [inline custom_completions.go:301]: I'm unsure about this one.  I believe it disables subcommand completion and `ValidArgs` completion for the non-interspersed case, right?

I think `ValidArgs` should still be done.
And I'm thinking subcommand completion should also be done. case like:
```
testprog --bool subcomm<TAB>
```
What do you think?

**Review by marckhouzam — COMMENTED** [review]
Nicely done!

File: custom_completions.go (hunk 2 above)
Comment (Luap99) [reply custom_completions.go]: You are correct this is wrong. It works fine without it. I added a test for it.

File: custom_completions_test.go (hunk 1 above)
Comment (Luap99) [reply custom_completions_test.go:1757]: Good catch this one was not working correct.

**Review by marckhouzam — COMMENTED** [review]
This is really nice.
I have mentioned in the comments two more problematic cases, but they are corner cases, especially the one about the `args` being incomplete.

I don't know if supporting these cases is worth the possible extra complexity to the code, but I haven't tried to fix it myself, so I'm not sure.  I personally would not consider them blockers as this PR already improves this scenario for the vast majority of cases.

Thanks for the continuing great work @Luap99 !

File: custom_completions_test.go [hunk 3]
```diff
@@ -1697,6 +1697,159 @@ func TestValidArgsFuncChildCmdsWithDesc(t *testing.T) {
...
+	if output != expected {
+		t.Errorf("expected: %q, got: %q", expected, output)
+	}
+
+	// Check that args are still completed even if flagname with ValidArgsFunction exists
+	output, err = executeCommand(rootCmd, ShellCompRequestCmd, "child", "--", "--string", "")
+	if err != nil {
+		t.Errorf("Unexpected error: %v", err)
+	}
+
```
[the remark this answers is not in the document]
Comment (Luap99) [reply custom_completions_test.go:1825]: Updated, the logic is definitely more complex now but I think it is still fine.

[the remark this answers is not in the document]
Comment (marckhouzam) [reply custom_completions_test.go:1825]: You did a good job, it is not that complex.

File: custom_completions_test.go [hunk 4]
```diff
@@ -1697,6 +1697,228 @@ func TestValidArgsFuncChildCmdsWithDesc(t *testing.T) {
...
+	if output != expected {
+		t.Errorf("expected: %q, got: %q", expected, output)
+	}
+
+	// Check that --validarg is added to args for the ValidArgsFunction
+	childCmd.ValidArgsFunction = func(cmd *Command, args []string, toComplete string) ([]string, ShellCompDirective) {
+		return args, ShellCompDirectiveDefault
+	}
+	output, err = executeCommand(rootCmd, ShellCompRequestCmd, "child", "--", "--validarg", "")
```
Comment (marckhouzam) [inline custom_completions_test.go:1888]: We have to use a valid flag for this test, which will then trigger the problem:
```
output, err = executeCommand(rootCmd, ShellCompRequestCmd, "child", "--", "--string", "")
```

`--validarg` is not found as a valid flag by `checkIfFlagCompletion()` so your new logic handling errors is triggered.  However, with `--string`, there will not be an error as it is a valid flag and the `--string` argument will be trimmed:
https://github.com/spf13/cobra/blob/a4ab3fa09e3d6c4ac3fb7deece97a910957305ab/custom_completions.go#L466

But this is such a rare (and possibly invalid) case that I don't think we should fix it (unless it is a very simple fix).

**Review by marckhouzam — APPROVED** [review]
This looks good to me.  It makes our completion support even more complete (no pun intended 😄 ).

File: custom_completions_test.go (hunk 4 above)
Comment (Luap99) [reply custom_completions_test.go:1888]: You are right catching these cases is more difficult and would require a bigger refactor of the `checkIfFlagCompletion` function.

**Comment by Luap99** [comment]
I just rebased to solve the merge conflict.

**Comment by marckhouzam** [comment]
@jpmcb  If you still have energy and time, this is another fix for completion which I had forgotten to mention.  This will make the last Fish shell test of the `cobra-completion-testing` repo work.

**Review by jpmcb — CHANGES_REQUESTED** [review]
Hi @Luap99 - getting a number of these PRs in to cut a release. Do you mind rebasing and resolving the conflict?

**Comment by Luap99** [comment]
@jpmcb done

**Review by jpmcb — APPROVED** [review]

**Pull request merged**